### DISCUSSION TITLE
Compare yaw optimization results SR and Scipy

### DISCUSSION THREAD

**Discussion by dhcho347** [NatLabRockies/floris#536] [Q&A] [answered]
There are differences in yaw optimization results between SR and Scipy above the rated wind speed.

In the case of the T0(turbine no1, rightmost one), 

there is no turbine either upstream nor downstream, 

so I don't know why it gives the 25degree yaw angle.

[simulation with NREL 5MW, 20ea WTGs with 700m spacing (4 *5 array)]

□ with wind speed 8m/s, it looks similar (under rated wind speed)
![image](https://user-images.githubusercontent.com/70368625/202483280-8c32777c-140a-401e-8675-0c157311330a.png)

□ with wind speed 14m/s, 15, 16, ... (over rated wind speed)
quite differ at no wake loss wind directions
![image](https://user-images.githubusercontent.com/70368625/202483806-1e39caee-c91f-4cc6-8f4e-0dddf481f87b.png)

□ df_opt results
![image](https://user-images.githubusercontent.com/70368625/202485002-611fa268-9718-4f39-8fd4-59ba63eedd47.png)

□ Visualization
![image](https://user-images.githubusercontent.com/70368625/202485063-7eb2be96-330e-4684-ab14-76b4715c6a11.png)

Thank you.

**Comment by Bartdoekemeijer** [comment]
The way yaw misalignment works in FLORIS is through an "effective" wind speed, rather than a direct manipulation of the turbine power production.

**For example:**
If you have a situation where `wind_speed = 14.0` and the turbine is yawed by `25.0` deg, then the effective wind speed becomes:
```
power_yawed = power_nonyaw * cos(yaw)^1.88
power_yawed = (0.5 * rho * wind_speed^3.0 * Ad) * cos(yaw)^1.88
power_yawed = (0.5 * rho * Ad) * (wind_speed^3.0 * cos(yaw)^1.88)
power_yawed = (0.5 * rho * Ad) * effective_wind_speed ^ 3.0
```
where `rho` is the air density in [kg/m3] and `Ad` is the rotor swept area in [m2]. By manipulating the equations we see:
```
effective_wind_speed =  wind_speed * cos(yaw)^(1.88 / 3)
```
You can find this equation in the code here: https://github.com/NREL/floris/blob/f450e3c44306a2d73cfe319a0816c36089215123/floris/simulation/turbine.py#L139
We use the turbine's effective inflow wind speed to interpolate the turbine power from the performance curve table (`cp`). So effectively, by increasing the yaw angle, the (effective) inflow wind speed that the turbine experiences and which is used for interpolation goes down, e.g., from 14 m/s to 13 m/s. In both scenarios, the turbine's power production should read 5.000 kW because in both scenarios the turbine is in _rated_ operation. However, due to numerical precision this may not be exactly right and the turbine power at 13 m/s (5000.001 kW) may be a very small number higher than the turbine power at 14 m/s (4999.999 kW). This is what is happening in your example.

This has been an issue before, see #340 and #385. This is why I added the option `verify_convergence` in the yaw optimizers, see:
https://github.com/NREL/floris/blob/f450e3c44306a2d73cfe319a0816c36089215123/floris/tools/optimization/yaw_optimization/yaw_optimization_base.py#L47
and
https://github.com/NREL/floris/blob/f450e3c44306a2d73cfe319a0816c36089215123/floris/tools/optimization/yaw_optimization/yaw_optimization_base.py#L519.

Edit: clarified my explanations, added some links to the code and other issues.

**Reply by dhcho347** [reply]
Hi, @Bartdoekemeijer,
Thank you for your quick response and detail explanation with previous discussion.

I've already checked class YawOptimization:, class YawOptimizationSR(YawOptimization):
your explanation exactly same with my understanding and expectation.

But, even with (verify_convergence=True), still I got 25 degree.
now I'm trying to find the way to get right results with SR 
.
.
.
**2nd:
There are 2 missing yaw angles in SR**

with default 
 - Ny_passes=[5, 4]
 - minimum_yaw_angle=0.0,  maximum_yaw_angle=25.0

First pass is okay.
![image](https://user-images.githubusercontent.com/70368625/202691314-0a6ba910-f430-4987-8ac5-a4870e6ebd74.png)


But Second pass has a problem.
![image](https://user-images.githubusercontent.com/70368625/202691346-47adf108-c1ce-49e5-ba0d-393cd0266375.png)


Due to the clip: The length of the first(0) and last(25) section has changed (6.25 > 3.125)
it will make the center does not match to first pass 
 - first and last: 1.5625, 23.4375 ( !~ 0, 25) 
 - the others: 6.25, 12.5, 18.75 ( ~ 6.25, 12.5,18.75)

https://github.com/NREL/floris/blob/5e71acc035b7cad6e76965b316caf0ca978a1957/floris/tools/optimization/yaw_optimization/yaw_optimizer_sr.py#L274-L278

results from ex10. there is no 23.4375 opt yaw degree
But, the others are there (21.875, 22.6563, 24.2188, 25)

When I print the values here
https://github.com/NREL/floris/blob/5e71acc035b7cad6e76965b316caf0ca978a1957/floris/tools/optimization/yaw_optimization/yaw_optimizer_sr.py#L199-L201

 c = 2, ids = [0,1,3,4],
- yaw_angles_subset = **[21.875, 22.6563, (x 23.4375), 24.2188, 25]** in last section
- 
![image](https://user-images.githubusercontent.com/70368625/202687187-fa7fb324-1417-4976-86e2-f68a87c23b2b.png)

- yaw_angles_subset = **[15.625, 17.1875, (x 18.75), 20.3125, 21.875]** in 4th section
- 
![image](https://user-images.githubusercontent.com/70368625/202688370-fc50c785-9ed0-4a7e-9250-7062cb7b4736.png)


18.75 is no problem, we already have that results from first pass,
but 23.4375, we don't have previous results, we just have 25 from first pass

I think we have to change the value in first and last section
for example in last section: we have to take 23.4375 instead of 25 


Please check these things.
Thank you

**Comment by dhcho347** [comment]
Hi, @Bartdoekemeijer 

I think this problem (above rated wind speed) was from the input of the power coefficient.

Because of decimal point cutting, fluctuations occur above the rated wind.
- google spread in tap [PowCurve]: 
  https://docs.google.com/spreadsheets/d/1NNB-2VbZFfed8HEw6xYvsuvMsQ1bM-K3LYQDsSkoFlA/edit#gid=31164577
- 5MW power curve graph in floris
![image](https://user-images.githubusercontent.com/70368625/203243828-b8467d3d-d24c-4286-b1f5-dbefeead1e91.png)
- making power curve with nrel_5MW.yaml
![image](https://user-images.githubusercontent.com/70368625/203244325-51138b00-6752-43ce-a427-722745bd9e76.png)

as you mentioned,
- power is 5,000,003.049 at 14m/s, 
- power is 4,999,997.283 at 13m/s

and this make these unwanted results
 - increase turbine and farm power with yaw angles. this is not common sense.
 - 1. below rated wind speed, the power should decrease [ following (cos(yaw angle))^1.88 ] (ex: 8m/s)
 - 2. quite above rated wind speed, the power should be same. (ex: 14m/s)
 - 3. around rated wind speed, same or decrease (ex: 12m/s)
![image](https://user-images.githubusercontent.com/70368625/203244771-ac96fb10-7d38-4fec-a829-66dfc49e9db8.png)

These are results that I want
with normal power curve (without fluctuation)

 - at 14m/s wind speed: same
![image](https://user-images.githubusercontent.com/70368625/203245980-2c8a371d-9204-4872-8c00-d17de8357e2d.png)

 - at 12m/s wind speed: decrease at specific yaw angle
![image](https://user-images.githubusercontent.com/70368625/203248287-bec34d2f-b09f-4244-84e0-49285679162f.png)


I temporarily got rid of the fluctuation above rated wind speed
with add these 6 lines in turbine.py
reset the values between rated and cut-out wind speed with ratedPower/1.225

https://github.com/dhcho347/floris/commit/4ffc45e9289477ba84599275903611f39d070ef0


and these are results from yaw optimization with SR
- after change
![image](https://user-images.githubusercontent.com/70368625/203247664-8d57e5d1-257b-499f-a03c-a79ec275a4f9.png)

- before change
![image](https://user-images.githubusercontent.com/70368625/203247687-08746a95-767f-4415-9d87-8cb092e8c7b0.png)

I run ex10 with before change.
- wind direction: 225 degrees
- wind speed: 13 ~ 25m/s
- then, the optimal yaw angles are keep changing from 0 to 25 degrees even without wind direction changes due to wrong power curve

 
Below rated wind
- after change
![image](https://user-images.githubusercontent.com/70368625/203251433-5aeec845-3543-4066-b158-bda43b968159.png)

- before change
![image](https://user-images.githubusercontent.com/70368625/203252036-5847872c-0fde-4f28-964a-b7aece72c08a.png)


There are differences in results.
It will be reviewed once more.

Please, check this out
Thank you.

**Comment by Bartdoekemeijer** [comment]
Hi @dhcho347. Thanks for sharing your findings! We found similar things when we looked into this a while ago. Your proposed solution does work, but it's hardcoded for specific turbines and under specific conditions, and hence not a solution we would want to adopt in FLORIS. Generally, it seems to me that there are two generic solutions to this issue:

1. Provide higher-resolution turbine power curves, e.g., in steps of 0.05 m/s.
2. Increase the default tolerance `in _verify_solutions_for_convergence`. Try increasing the default value `min_power_gain_for_yaw` from 0.02 to 0.1 or higher in https://github.com/NREL/floris/blob/f450e3c44306a2d73cfe319a0816c36089215123/floris/tools/optimization/yaw_optimization/yaw_optimization_base.py#L524
This would essentially require any yaw offset to generate at least a certain percentage of power uplift. If this uplift is numerically insignificant, that yaw offset should be nullified.

A second consideration is that we also don't optimize the yaw angles very often at 14 m/s wind speeds. Very often, we don't apply yaw offsets above a certain wind speed (typically near region 2.5) due to load considerations. In commercial application, we typically ramp up near cut-in and ramp down near rated wind speeds to prevent unnecessary yaw offsets and also prevent damage due to misalignment at high wind speeds. For example, we ramp up the yaw offset from 0 deg at 4 m/s to 20 deg at 5 m/s, then hold this maximum offset until 10 m/s, and then ramp down the yaw offset to 0 deg at 12 m/s.

**Comment by dhcho347** [comment]
Hi, @Bartdoekemeijer
Thanks for quick reply and informing some generic solutions.

Currently, I think the problem is relating with input data not algorithm.
so, I prefer option1 than option2.
and what about input a power curve instead of Cp like as Windpro(EMD). 
We can calculate Cp from power curve easily.

I completely agree with the opinion that yaw results above rated wind speed are not important now.
I want to use yaw steering idea to our company's wind farm.
So, one of my consortiums is doing time domain calculation(FloriDyn, DTU) mainly with below rated wind speed yaw results as same reason with your explanation.
and they are finding best period of moving average, limit and rate for wind directions and so on.
After that, we will check and compare the extreme and fatigue loads are still stay in the envelop of design cert. by using GHbladed(with updated control dll) on component level.

Get back to the point,
I'm just looking at it because I just want to see what makes these wrong results. Due to curiousity rather than necessity.

Today I don't have a time to see, so there is nothing new one
but I'd like to do something like you were giving me quick answer. 

I just share my previous testing data a bit more.
 1.1 with more higher accuracy intputs for Cp
 - input
![image](https://user-images.githubusercontent.com/70368625/203496495-4d50bd32-d284-429d-812c-162fcde00b42.png)

 - results: It was work, but not enough
![image](https://user-images.githubusercontent.com/70368625/203496182-2e42d9b0-8149-4360-b9f8-c8ebcf6041e7.png)

 1.2 with 0.05m/s steps inputs for Cp
 - input
 
![image](https://user-images.githubusercontent.com/70368625/203496980-b2b23ab4-16d7-4956-9453-810d26f0829b.png)

 - results: it will be similar with 1.1 because it still has fluctuation.

 1.3 input with power curve instead of Cp (without fluctuation)
 - results: it will be same with hardcode

and please see this discussion, too
when you have a time, #539 

after I find new things later, I will discuss more.
Thank you for your help and sharing your experience and knowledge.
Have a nice day.

**Comment by dhcho347** [comment]
Hi, @Bartdoekemeijer 

https://docs.google.com/spreadsheets/d/1dqvuC4zRREso7t41v-PFCzimFyxLcmUcwxzB36RhFAE/edit#gid=1610790205

This spread is same procedure as Floris
from velocities to power for easy viewing at what stage a problem occurs.

There are false power increase even with lower effective wind speed due to wrongly generated power curve from truncated Cp

The false power increase even with low effective wind speed
gives wrong yaw angle results.

**Comment by paulf81** [comment, marked as answer]
With PR #590  we have hopefully resolved issues related to fluctuations above rated, can you see if this helped @dhcho347 ?

**Reply by dhcho347** [reply]
Hi, @paulf81 
This is the general solution I was looking for

now, I can change my temporary measure to yours. (inner_power in turbine.py)

https://github.com/dhcho347/floris/commit/4ffc45e9289477ba84599275903611f39d070ef0#diff-8e60d53418d5e034e6d512b76142d763898ca31f03b68b4b1da1b9be04e46c1fL383-L384
```python
       inner_power = 0.5 * self.rotor_area * self.fCp_interp(wind_speeds) * self.generator_efficiency * wind_speeds ** 3

        if self.turbine_type =='nrel_5MW':
            inner_power[ (wind_speeds>=11.4) * (wind_speeds<=25.0)]=5000000/1.225
        if self.turbine_type =='iea_10MW':
            inner_power[ (wind_speeds>=11) * (wind_speeds<=25.0)]=10000000/1.225
```

These were test results.
![image](https://user-images.githubusercontent.com/70368625/233255942-367c3cfd-534e-4639-8319-c6fe7fed357c.png)


and, I think there are still a few things left to consider
 - 225 degree is good 
 - 226~229 degrees... I would like to see these results one more time.
![image](https://user-images.githubusercontent.com/70368625/233256246-9fcf4645-8a83-4d55-a3a8-a86b377ff7ca.png)
